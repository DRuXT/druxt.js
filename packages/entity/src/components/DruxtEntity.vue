<script>
import merge from 'deepmerge'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
import { mapActions } from 'vuex'

/**
 * The `<DruxtEntity />` component fetches the Entity JSON:API resource and
 * schema data to render the Entities fields based on the Drupal Display
 * Mode configuration.
 * 
 * Features:
 * - Display mode based field system
 * - Scoped slots
 * - Query settings
 *
 * @example @lang vue
 * <DruxtEntity
 *   type="node--article"
 *   :uuid="uuid"
 *   mode="teaser"
 * />
 */
export default {
  name: 'DruxtEntity',

  extends: DruxtModule,

  /** */
  props: {
    /**
     * Drupal display mode.
     *
     * @type {string}
     * @default default
     */
    mode: {
      type: String,
      default: 'default'
    },

    /**
     * JSON:API resource type.
     *
     * @type {string}
     */
    type: {
      type: String,
      required: true
    },

    /**
     * Entity UUID.
     *
     * @type {string}
     */
    uuid: {
      type: [Boolean, String],
      default: false
    },

    /**
     * Drupal display schema type, 'view' or 'form'.
     * 
     * @type {('view'|'form')}
     */
    schemaType: {
      type: String,
      default: undefined,
    },
  },

  /**
   * Nuxt.js fetch method.
   */
  async fetch() {
    // Fetch Schema.
    try {
      this.schema = await this.getSchema({
        resourceType: this.type,
        mode: this.mode,
        schemaType: this.schemaType || 'view',
      })
    } catch(e) {}

    // Build wrapper component object.
    const options = this.getModuleComponents()
    let component = {
      is: (((options.filter(o => o.global) || [])[0] || {}).name || 'DruxtWrapper'),
      options: options.map(o => o.name) || [],
    }

    // Get wrapper component data to merge with module settings.
    const wrapperData = await this.getWrapperData(component.is)
    component.settings = merge((this.$druxtEntity || {}).options || {}, wrapperData.druxt || {}, { arrayMerge: (dest, src) => src })

    // Fetch Entity resource.
    if (this.uuid && !this.value) {
      const query = this.getQuery(component.settings)
      const entity = (await this.getResource({ type: this.type, id: this.uuid, query })).data
      this.model = JSON.parse(JSON.stringify(entity || {}))
    }

    // Get scoped slots.
    component.slots = Object.keys(this.getScopedSlots())

    // Build wrapper component propsData.
    component = { ...component, ...this.getModulePropsData(wrapperData.props) }

    // Set component data.
    this.component = component
  },

  fetchKey(getCounter) {
    const parts = ['DruxtEntity', this.type, this.uuid, this.mode, this.schemaType].filter((o) => o)
    return [...parts, getCounter(parts.join(':'))].join(':')
  },

  /**
   * @property {object} model - The model object.
   * @property {object} schema - The DruxtSchema object.
   */
  data: ({ type, value }) => ({
    model: {
      attributes: {},
      relationships: {},
      type,
      ...value,
    },
    schema: null,
  }),

  /** */
  computed: {
    /**
     * The Entity object.
     *
     * @return {object}.
     */
    entity: ({ model }) => ({ ...model }),

    /**
     * Entity fields based on Display mode.
     *
     * @return {object}
     */
    fields: ({ errors, isEmpty, model, schema, schemaType }) => {
      if (!schema) return false

      const fields = {}
      for (const field of schema.fields) {
        const relationship = !!((field.settings || {}).storage || {}).target_type || !!(model.relationships || {})[field.id]
        const value = relationship ? ((model || {}).relationships || {})[field.id] : ((model || {}).attributes || {})[field.id]

        // Filter out empty fields if not using the Form schema type.
        // @todo - Make this configurable?
        if (schemaType !== 'form' && isEmpty(value)) continue

        fields[field.id] = {
          id: field.id,
          // @todo - Remove deprecated 'data'.
          data: value,
          errors: (errors || []).filter((o) => ((o.source || {}).pointer || '').startsWith(`/data/attributes/${field.id}`)),
          relationship,
          schema: {
            config: schema.config,
            ...field,
          },
          value,
        }
      }

      return fields
    },
  },

  watch: {
    mode() {
      this.$fetch()
    },

    schemaType() {
      this.$fetch()
    },

    type() {
      this.$fetch()
    },

    uuid() {
      this.$fetch()
    },
  },

  methods: {
    /**
     * Builds the query for the JSON:API request.
     *
     * @param {ModuleSettings} settings - The merged module and component settings object.
     *
     * @return {boolean|object}
     */
    getQuery(settings) {
      const query = new DrupalJsonApiParams()

      // Build fields list.
      const fields = (settings.query || {}).schema
        ? [...((this.schema || {}).fields || []).map((o) => o.id), ...((settings.query || {}).fields || [])]
        : ((settings.query || {}).fields || [])
      if (fields.length) {
        query.addFields(this.type, fields)
        return query
      }
      return false
    },

    /**
     * Checks if an Entity field is empty.
     *
     * @param {*} value - Field value.
     * @return {boolean}
     */
    isEmpty(value) {
      if (typeof value === 'undefined') return true

      if (!value) return true

      if (Array.isArray(value.data) && !value.data.length) return true

      if (typeof value.data !== 'undefined' && !value.data) return true

      return false
    },

    /**
     * Maps Vuex action to methods.
     */
    ...mapActions({
      getResource: 'druxt/getResource',
      getSchema: 'druxtSchema/get'
    })
  },

  /** DruxtModule settings */
  druxt: {
    /**
     * Provides the available component naming options for the Druxt Wrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {ComponentOptions}
     */
    componentOptions: ({ mode, schema, schemaType, type }) => ([
      [(schema || {}).resourceType || type, ((schema || {}).config || {}).mode || mode, ((schema || {}).config || {}).schemaType || schemaType || 'view'],
      [((schema || {}).config || {}).mode || mode],
    ]),

    /**
     * Provides propsData for the DruxtWrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {PropsData}
     */
    propsData: ({ fields, model, schema }) => ({ entity: model, fields, schema, value: model }),

    /**
     * Provides the scoped slots object for the Module render function.
     *
     * A scoped slot is provided for each field being rendered, as per the
     * current display mode.
     * 
     * Additionally, the `default` slot will render all fields as per the
     *
     * @example <caption>DruxtEntity**ResourceType**.vue</caption> @lang vue
     * <template>
     *   <div>
     *     <slot name="content" />
     *     <slot :name="field_name" />
     *   </div>
     * </template>
     *
     * @return {ScopedSlots} The Scoped slots object.
     */
    slots(h) {
      const scopedSlots = {}

      if (!this.schema) {
        if (this.$nuxt.context.isDev) {
          scopedSlots.default = (attrs) => h('details',
            {
              attrs,
              style: {
                border: '2px dashed lightgrey',
                margin: '0.5em 0',
                padding: '1em',
              },
            },
            [
              h('summary', [`[DruxtEntity] Missing schema for '${this.type}--${this.mode}'`]),
              h('br'),
              h('label', ['Component options:', h('ul', this.component.options.map((s) => h('li', [s])))]),
              h('br'),
              h('label', ['Entity:', h('pre', [JSON.stringify(this.entity, null, '  ')])]),
            ]
          )
        }
        return scopedSlots
      }

      // Build scoped slots for each field.
      Object.entries(this.fields).map(([id, field]) => {
        scopedSlots[id] = (attrs) => h('DruxtField', {
          attrs,
          key: id,
          props: field,
          on: {
            input: (value) => {
              const type = !field.relationship ? 'attributes' : 'relationships'
              this.model[type][id] = value
              this.$emit('input', this.model)
            }
          },
          ref: id,
        })
      })

      // Build default slot.
      scopedSlots.default = (attrs) => Object.entries(this.fields).map(([id]) => scopedSlots[id](attrs))

      return scopedSlots
    },
  },
}

/**
 * Provides the available naming options for the Wrapper component.
 *
 * @typedef {array[]} ComponentOptions
 *
 * @example @lang js
 * [
 *   'DruxtEntity[ResourceType][DisplayMode][SchemaType]',
 *   'DruxtEntity[ResourceType][DisplayMode]',
 *   'DruxtEntity[ResourceType]',
 *   'DruxtEntity[DisplayMode]',
 * ]
 *
 * @example <caption>Article Node (default)</caption> @lang js
 * [
 *   'DruxtEntityNodeArticleDefaultView',
 *   'DruxtEntityNodeArticleDefault',
 *   'DruxtEntityNodeArticle',
 *   'DruxtEntityNodeDefault',
 * ]
 */

/**
 * Provides settings for the Entity module, via the `nuxt.config.js` `druxt.entity`
 * or the Wrapper component `druxt` object.
 *
 * @typedef {object} ModuleSettings
 * @param {string[]} fields - An array of fields to filter from the JSON:API Resource.
 * @param {boolean} schema - Whether to automatically detect fields to filter, per the Display mode.
 *
 * @example @lang js
 * {
 *   fields: [],
 *   schema: false,
 * }
 *
 * @example @lang vue
 * <script>
 * export default {
 *   druxt: {
 *     query: {
 *       fields: ['title']
 *       schema: true,
 *     },
 *   }
 * }
 */

/**
 * Provides propsData for use in the Wrapper component.
 *
 * @typedef {object} PropsData
 * @param {object} entity - The Drupal Entity JSON:API resource data.
 * @param {object} fields - Drupal Entity Fields data.
 * @param {object} schema - DruxtJS Schema object.
 * @param {object} value - The Entity value.
 *
 * @example @lang js
 * {
 *   entity: {
 *     attributes: {},
 *     id: '43118086-cca5-4c62-b11e-f1d870050ebd',
 *     links: {},
 *     relationships: {},
 *     type: 'node--article',
 *   },
 *   fields: {
 *     body: {},
 *     field_media_image: {},
 *     field_tags: {},
 *   },
 *   schema: {
 *     config: {},
 *     fields: {},
 *     groups: {},
 *     id: 'node--article--default--view',
 *     resourceType: 'node--article',
 *   },
 *   value: {},
 * }
 */

/**
 * Provides scoped slots for use in the Wrapper component.
 *
 * @typedef {object} ScopedSlots
 * @param {function} * - Slot per field.
 * @param {function} default - All fields per Display mode.
 *
 * @example <caption>DruxtEntity**ResourceType**.vue</caption> @lang vue
 * <template>
 *   <div>
 *     <slot name="content" />
 *     <slot :name="field_name" />
 *   </div>
 * </template>
 */
</script>
