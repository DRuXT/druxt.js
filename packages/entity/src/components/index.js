/**
 * The `<DruxtEntity />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntity
 * @see {@link ./DruxtEntity|DruxtEntity}
 */
export { default as DruxtEntity } from './DruxtEntity.vue'

/**
 * The `<DruxtEntityForm />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntityForm
 * @see {@link ./DruxtEntityForm|DruxtEntityForm}
 */
export { default as DruxtEntityForm } from './DruxtEntityForm.vue'

/**
 * The `<DruxtEntityFormButtons />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntityFormButtons
 * @see {@link ./DruxtEntityFormButtons|DruxtEntityFormButtons}
 */
export { default as DruxtEntityFormButtons } from './DruxtEntityFormButtons.vue'

/**
 * The `<DruxtField />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtField
 * @see {@link ./DruxtField|DruxtField}
 */
export { default as DruxtField } from './DruxtField.vue'

/**
 * The `<DruxtRouterEntity />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtRouterEntity
 * @see {@link ./DruxtRouterEntity|DruxtRouterEntity}
 */
export { default as DruxtRouterEntity } from './DruxtRouterEntity.vue'

export * from './fields'
