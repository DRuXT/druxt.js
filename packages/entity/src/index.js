/**
 * Vue.js mixins.
 *
 * @type {object}
 * @exports DruxtEntityMixins
 * @see {@link ./mixins|DruxtEntityMixins}
 * @see {@link ./mixins/entity|DruxtEntityMixin}
 * @see {@link ./mixins/field|DruxtFieldMixin}
 */
export * from './mixins'

/**
 * Default function to alert user to incorrectly installed module.
 *
 * This was added as part of the @nuxt/kit update due to breaking changes.
 */
export default () => {
  throw new Error("DruxtEntity Nuxt module must be installed as 'druxt-entity/nuxt'")
}
