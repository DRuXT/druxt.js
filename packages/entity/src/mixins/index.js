export * from './context.js'
export * from './componentSuggestion.js'

/**
 * Vue.js mixin for DruxtEntity and DruxtEntityForm Wrapper components.
 *
 * @type {object}
 * @exports DruxtEntityMixin
 * @see {@link ./entity|DruxtEntityMixin}
 */
export { DruxtEntityMixin } from './entity'

/**
 * Vue.js mixin for DruxtField Wrapper components.
 *
 * @type {object}
 * @exports DruxtFieldMixin
 * @see {@link ./field|DruxtFieldMixin}
 */
export { DruxtFieldMixin } from './field'
