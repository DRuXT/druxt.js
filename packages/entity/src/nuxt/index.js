import { defineNuxtModule, installModule } from '@nuxt/kit'
import { join } from 'path'
// import DruxtEntityStorybook from '../nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds dependant modules.
 * - Adds Nuxt plugin.
 * - Adds Nuxt Storybook integration.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-entity'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @todo Document options.
 *
 * @property {object} options.druxt - DruxtJS root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 */
const DruxtEntityNuxtModule = defineNuxtModule({
  meta: {
    name: 'druxt-entity',
  },
  defaults: {
    baseUrl: '',
    endpoint: '/jsonapi'
  },

  async setup(moduleOptions, nuxt) {
    // Set default options.
    const options = {
      baseUrl: moduleOptions.baseUrl,
      ...nuxt.options?.druxt,
      entity: {
        query: {},
        ...nuxt.options?.druxt?.entity,
        ...moduleOptions,
        components: {
          fields: false,
          ...nuxt.options?.druxt?.entity?.components,
          ...moduleOptions?.components
        }
      }
    }

    // Add dependant modules.
    await installModule('druxt/nuxt', options, nuxt)
    await installModule('druxt-schema/nuxt', { baseUrl: options.baseUrl }, nuxt)

    // Register components directories.
    nuxt.hook('components:dirs', dirs => {
      dirs.push({
        path: join(__dirname, '../dist/components'),
        ignore: ['fields']
      })
      if (options.entity.components.fields) {
        dirs.push({ path: join(__dirname, '../dist/components/fields') })
      }
    })

    // Nuxt Storybook.
    // @TODO - @nuxt/kit and @nuxt/storybook aren't compatible.
    // this.nuxt.hook('storybook:config', async ({ stories }) => {
    //   await DruxtEntityStorybook.call(this, { stories })
    // })
  }
})

export default DruxtEntityNuxtModule
