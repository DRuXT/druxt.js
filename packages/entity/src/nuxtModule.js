import { resolve } from 'path'
import DruxtEntityStorybook from './nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds dependant modules.
 * - Adds Nuxt plugin.
 * - Adds Nuxt Storybook integration.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-entity'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @todo Add module level options.
 * @todo Document options.
 *
 * @property {object} options.druxt - DruxtJS root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 */
const DruxtEntityNuxtModule = function (moduleOptions = {}) {
  const { options } = this

  // Use root level Druxt options.
  if (typeof options === 'undefined' || !options.druxt) {
    throw new TypeError('Druxt settings missing.')
  }

  // Add dependant modules.
  const modules = ['druxt', 'druxt-schema'].filter((module) => !options.modules.includes(module))
  for (const module of modules) {
    this.addModule(module)
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/plugin.js'),
    fileName: 'druxt-entity.js',
    options: options.druxt
  })

  // Nuxt Storybook.
  this.nuxt.hook('storybook:config', async ({ stories }) => {
    await DruxtEntityStorybook.call(this, { stories })
  })
}

DruxtEntityNuxtModule.meta = require('../package.json')

export { DruxtEntityNuxtModule }
