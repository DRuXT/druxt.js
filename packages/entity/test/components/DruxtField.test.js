import 'regenerator-runtime/runtime'
import { createLocalVue, mount } from '@vue/test-utils'
import { getMockResource } from 'druxt-test-utils'
import Vuex from 'vuex'
import mockAxios from 'jest-mock-axios'

import { DruxtClient, DruxtStore } from '../../../druxt/src'
import { DruxtSchemaStore } from '../../../schema/src'
import DruxtEntity from '../../src/components/DruxtEntity.vue'
import DruxtField from '../../src/components/DruxtField.vue'

jest.mock('axios')

// Setup local vue instance.
const localVue = createLocalVue()
localVue.component('DruxtEntity', DruxtEntity)
localVue.use(Vuex)

let store

const mocks = {
  $fetchState: {
    pending: false
  },
  $nuxt: {
    context: {
      isDev: false,
    }
  },
}

const mountComponent = async ({ data, entity, field, mode = 'default', options = {}, uuid, schema }) => {
  data = data || { ...entity.attributes, ...entity.relationships }  

  if (!schema) {
    schema = require(`../../../../test/__fixtures__/schemas/${entity.type}--${mode}--view.json`)
  }

  const fieldSchema = schema.fields.find(element => element.id === field)

  const propsData = {
    data: data[field],
    schema: {
      config: schema.config,
      ...fieldSchema,
    },
    value: data[field],
    relationship: !!entity.relationships[field]
  }

  return mount(DruxtField, { ...options, localVue, mocks, propsData, store })
}

describe('DruxtField', () => {
  beforeEach(() => {
    mockAxios.reset()

    // Setup vuex store.
    store = new Vuex.Store()

    DruxtStore({ store })
    store.$druxt = new DruxtClient('https://demo-api.druxtjs.org')

    DruxtSchemaStore({ store })
    store.$druxtSchema = {
      import: (schema) => {
        return require(`../../../../test/__fixtures__/schemas/${schema}.json`)
      }
    }

    store.app = { context: { error: jest.fn() }, store }
  })

  test('defaults', async () => {
    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'body',
      uuid: mockPage.data.id,
    })

    // Props.
    expect(wrapper.vm.errors).toStrictEqual([])
    expect(wrapper.vm.relationship).toBe(false)
    expect(wrapper.vm.schema).toStrictEqual(expect.any(Object))
    expect(wrapper.vm.options).toStrictEqual({})
    expect(wrapper.vm.value).toStrictEqual(expect.any(Object))

    // Fetch key.
    expect(DruxtField.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtField:body:0')

    // Data.
    expect(wrapper.vm.model).toStrictEqual(wrapper.vm.value)

    // Computed.
    expect(wrapper.vm.data).toStrictEqual(wrapper.vm.value)
    expect(wrapper.vm.isBoolean).toStrictEqual(false)
    expect(wrapper.vm.isDateTime).toStrictEqual(false)
    expect(wrapper.vm.isFile).toStrictEqual(false)
    expect(wrapper.vm.isImage).toStrictEqual(false)
    expect(wrapper.vm.isLink).toStrictEqual(false)
    expect(wrapper.vm.isMultiple).toStrictEqual(false)
    expect(wrapper.vm.isText).toStrictEqual(false)

    // DruxtModule
    expect(wrapper.vm.component).toStrictEqual({
      $attrs: {},
      is: 'DruxtWrapper',
      options: [],
      props: {},
      propsData: {},
      settings: {},
      slots: [],
    })

    // Slots.
    expect(Object.keys(wrapper.vm.getScopedSlots())).toStrictEqual([
      'label',
      'field-0',
      'default',
    ])

    const h = jest.fn()
    const slotsMock = {
      ...mocks,
      component: {
        options: ['DruxtFieldTest'],
      },
      label: {
        position: 'above',
        text: 'Label',
      },
      schema: {
        label: {
          text: 'Label',
        },
      },
    }
    expect(Object.keys(DruxtField.druxt.slots.call(slotsMock, h))).toStrictEqual([
      'label',
      'label-above',
      'field-0',
      'default',
    ])

    slotsMock.label.position = 'inline'
    slotsMock.$nuxt.context.isDev = true
    const slots = DruxtField.druxt.slots.call(slotsMock, h)
    expect(Object.keys(slots)).toStrictEqual([
      'label',
      'label-inline',
      'field-0',
      'default'
    ])
    slots.default({})
  })

  test('body - view', async () => {
    // Register custom wrapper component.
    localVue.component('DruxtFieldTextDefault', {
      render(h) {
        return h('slot')
      }
    })

    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'body',
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtField.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtField:body:0')

    // Computed props.
    expect(wrapper.vm.data).toStrictEqual(wrapper.vm.model)
    expect(wrapper.vm.isBoolean).toStrictEqual(false)
    expect(wrapper.vm.isDateTime).toStrictEqual(false)
    expect(wrapper.vm.isFile).toStrictEqual(false)
    expect(wrapper.vm.isImage).toStrictEqual(false)
    expect(wrapper.vm.isLink).toStrictEqual(false)
    expect(wrapper.vm.isMultiple).toStrictEqual(false)
    expect(wrapper.vm.isText).toStrictEqual(false)

    // DruxtModule.
    expect(wrapper.vm.component.is).toBe('DruxtFieldTextDefault')
    expect(wrapper.vm.component.options).toStrictEqual([
      'DruxtFieldTextDefaultBodyView',
      'DruxtFieldTextDefaultBody',
      'DruxtFieldTextDefaultView',
      'DruxtFieldDefaultView',
      'DruxtFieldTextDefault',
      'DruxtFieldDefault',
    ])
    expect(Object.keys(wrapper.vm.value)).toStrictEqual(['value', 'format', 'processed', 'summary'])
  })

  test('boolean - form', async () => {
    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'status',
      schema: {
        config: { schemaType: 'form' },
        fields: [{
          id: 'status',
          type: 'boolean_checkbox',
        }],
      },
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Rendered element.
    expect(wrapper.vm.isBoolean).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()

    // Input.
    const input = wrapper.find('input[type="checkbox"]')
    expect(input.exists()).toBe(true)
    expect(wrapper.vm.model).toBe(true)
    input.trigger('click')
    // expect(wrapper.vm.model).toBe(false)
  })

  test('datetime - form', async () => {
    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'created',
      schema: {
        config: { schemaType: 'form' },
        fields: [{
          id: 'created',
          settings: { display: {} },
          type: 'datetime_timestamp',
        }],
      },
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Rendered element.
    expect(wrapper.vm.isDateTime).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()

    // Input.
    const input = wrapper.find('input[type="datetime-local"]')
    expect(input.exists()).toBe(true)
    expect(input.element.value).toBe('2021-05-01T06:24:09.000')
    input.element.value = '2021-07-03T16:20'
    input.trigger('input')
    expect(wrapper.vm.model).toBe('2021-07-03T16:20+00:00') 
  })

  test('file - view', async () => {
    const mockImage = await getMockResource('media--image')
    const wrapper = await mountComponent({
      components: { DruxtEntity },
      entity: mockImage.data,
      field: 'field_media_image',
      schema: {
        config: { schemaType: 'view' },
        fields: [{
          id: 'field_media_image',
          settings: { display: {} },
          type: 'file_default',
        }],
      },
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)
    const entity = wrapper.findComponent(DruxtEntity)
    expect(entity.exists()).toBe(true)
    await entity.vm.$options.fetch.call(entity.vm)

    // Rendered element.
    expect(wrapper.vm.isFile).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('image - view', async () => {
    const mockImage = await getMockResource('media--image')
    const wrapper = await mountComponent({
      components: { DruxtEntity },
      entity: mockImage.data,
      field: 'field_media_image',
      schema: {
        config: { schemaType: 'view' },
        fields: [{
          id: 'field_media_image',
          settings: { display: {} },
          type: 'image',
        }],
      },
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)
    const entity = wrapper.findComponent(DruxtEntity)
    expect(entity.exists()).toBe(true)
    await entity.vm.$options.fetch.call(entity.vm)

    // Rendered element.
    expect(wrapper.vm.isImage).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('link - view', async () => {
    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'field_link',
      schema: {
        config: { schemaType: 'view' },
        fields: [{
          id: 'field_link',
          settings: { display: {} },
          type: 'link',
        }],
      },
      data: { field_link: {
        title: 'DruxtJS',
        uri: 'https://druxtjs.org',
      }},
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Rendered element.
    expect(wrapper.vm.isLink).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('text - form', async () => {
    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'title',
      schema: {
        config: { schemaType: 'form' },
        fields: [{
          id: 'title',
          settings: { display: {} },
          type: 'string_textfield',
        }],
      },
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Rendered element.
    expect(wrapper.vm.isText).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()

    // Input.
    const input = wrapper.find('input')
    expect(input.exists()).toBe(true)
    expect(wrapper.vm.model).toBe('About Umami')
    input.element.value = 'Test'
    input.trigger('input')
    expect(wrapper.vm.model).toBe('Test')
  })

  test('relationship - view', async () => {
    const mockArticle = await getMockResource('node--article')
    const wrapper = await mountComponent({
      entity: mockArticle.data,
      field: 'field_media_image',
      mode: 'card',
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtField.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtField:field_media_image:0')

    // Computed props.
    expect(wrapper.vm.data).toStrictEqual(wrapper.vm.model)
    expect(wrapper.vm.isBoolean).toStrictEqual(false)
    expect(wrapper.vm.isDateTime).toStrictEqual(false)
    expect(wrapper.vm.isFile).toStrictEqual(false)
    expect(wrapper.vm.isImage).toStrictEqual(false)
    expect(wrapper.vm.isLink).toStrictEqual(false)
    expect(wrapper.vm.isMultiple).toStrictEqual(false)
    expect(wrapper.vm.isText).toStrictEqual(false)

    // DruxtModule.
    expect(wrapper.vm.component.is).toBe('DruxtWrapper')
    expect(wrapper.vm.component.options).toStrictEqual([
      'DruxtFieldEntityReferenceEntityViewFieldMediaImageView',
      'DruxtFieldEntityReferenceEntityViewFieldMediaImage',
      'DruxtFieldEntityReferenceEntityViewView',
      'DruxtFieldDefaultView',
      'DruxtFieldEntityReferenceEntityView',
      'DruxtFieldDefault',
    ])
    expect(Object.keys(wrapper.vm.value)).toStrictEqual(['data', 'links'])
  })

  test('fallback - form', async () => {
    const mockPage = await getMockResource('node--page')
    const wrapper = await mountComponent({
      entity: mockPage.data,
      field: 'path',
      schema: {
        config: { schemaType: 'form' },
        fields: [{
          id: 'path',
          settings: { display: {} },
        }],
      },
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Rendered element.
    expect(wrapper.html()).toMatchSnapshot()

    // Input.
    const textarea = wrapper.find('textarea')
    expect(textarea.exists()).toBe(true)
    expect(wrapper.vm.model).toStrictEqual({
      alias: '/about-umami',
      langcode: 'en',
      pid: 101,
    })
    textarea.element.value = 'Test'
    textarea.trigger('input')
    expect(wrapper.vm.model).toBe('Test')
  })
})
