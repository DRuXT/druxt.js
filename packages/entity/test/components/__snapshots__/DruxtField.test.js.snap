// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtField boolean - form 1`] = `
"<div data=\\"true\\">
  <div data=\\"true\\" errors=\\"\\" schema=\\"[object Object]\\" value=\\"true\\"><label><strong>Status:</strong><br>
      <div><input type=\\"checkbox\\"></div></label></div>
</div>"
`;

exports[`DruxtField datetime - form 1`] = `
"<div data=\\"2021-05-01T06:24:09+00:00\\">
  <div data=\\"2021-05-01T06:24:09+00:00\\" errors=\\"\\" schema=\\"[object Object]\\" value=\\"2021-05-01T06:24:09+00:00\\"><label><strong>Created:</strong><br>
      <div><input type=\\"datetime-local\\"></div></label></div>
</div>"
`;

exports[`DruxtField fallback - form 1`] = `
"<div data=\\"[object Object]\\">
  <div data=\\"[object Object]\\" errors=\\"\\" schema=\\"[object Object]\\" value=\\"[object Object]\\"><label><strong>Path:</strong><br>
      <div><textarea></textarea></div></label></div>
</div>"
`;

exports[`DruxtField file - view 1`] = `
"<div data=\\"[object Object]\\">
  <div data=\\"[object Object]\\" errors=\\"\\" relationship=\\"true\\" schema=\\"[object Object]\\" value=\\"[object Object]\\">
    <div>
      <div entity=\\"[object Object]\\" value=\\"[object Object]\\"><a target=\\"_blank\\" href=\\"/sites/default/files/mediterranean-quiche-umami.jpg\\">mediterranean-quiche-umami.jpg</a></div>
    </div>
  </div>
</div>"
`;

exports[`DruxtField image - view 1`] = `
"<div data=\\"[object Object]\\">
  <div data=\\"[object Object]\\" errors=\\"\\" relationship=\\"true\\" schema=\\"[object Object]\\" value=\\"[object Object]\\">
    <div>
      <div entity=\\"[object Object]\\" value=\\"[object Object]\\"><img src=\\"/sites/default/files/mediterranean-quiche-umami.jpg\\"></div>
    </div>
  </div>
</div>"
`;

exports[`DruxtField link - view 1`] = `
"<div data=\\"[object Object]\\">
  <div data=\\"[object Object]\\" errors=\\"\\" schema=\\"[object Object]\\" value=\\"[object Object]\\"><a href=\\"https://druxtjs.org\\" target=\\"_blank\\">DruxtJS</a></div>
</div>"
`;

exports[`DruxtField text - form 1`] = `
"<div data=\\"About Umami\\">
  <div data=\\"About Umami\\" errors=\\"\\" schema=\\"[object Object]\\" value=\\"About Umami\\"><label><strong>Title:</strong><br>
      <div><input></div></label></div>
</div>"
`;
